Python/chapter01/1.9 - String Rotation/miguel_1.9_sol.py

```diff
@@ -0,0 +1,100 @@
+"""
+Python version 3.7.0
+1.9 - String Rotation
+Assume you have a method isSubstring which
+checks if one word is a substring
+of another.  Given two strings, s1 and s2,
+write code to check if s2 is a
+rotation of s1 using only one call to isSubstring
+(e.g. 'waterbottle' is a rotation of 'erbottlewat')
+"""
+import unittest
+
+
+def is_substring(sub: str, s: str) -> bool:
+    """
+    is_substring checks if 'sub' is a substring of 's'
+    :param sub: the substring
+    :param s: the alleged full string
+    :return: true if sub is a substring of s, false otherwise
+    """
+    return sub in s
+
+
+def string_rotation(s1: str, s2: str) -> bool:
+    """
+    Given two strings, string_rotation will check if s2 is a rotation of s1
+    using only one call to isSubstring.
+    We will loop through the original string and compare chars of
+    rotated string.  At the first character equality, we will
+    compare the rest of the original string to the rotated
+    string's substring of the same length.
+    If they match, then we save the index of the
+    rotated string + 1, and if whatever is left of
+    s1 is a substring of s2, then we must have a
+    rotation since we compared the complementary part already.
+    Runtime: Worst case:  O(n^2)
+    Space Complexity:  O(1)
+    :param s1: the rotated string
+    :param s2: the original string
+    :return: true if s2 is a rotation of s1, false otherwise
+    """
+    # not rotation if lengths don't match
+    if len(s1) != len(s2):
+        return False
+    # if they match, but both empty, return true
+    if s1 == '' and s2 == '':
+        return True
+    size = len(s1)
+    start_idx = -1
+    for i, c in enumerate(s2):
```

Comment:
you are circumventing the restriction of doing a single call to `is_substring` by kinda sorta checking for a partial substring by hand, which goes against the spirit of the problem.